prov/efa/src/rdm/efa_rdm_rma.c

```diff
@@ -197,24 +197,25 @@ ssize_t efa_rdm_rma_readmsg(struct fid_ep *ep, const struct fi_msg_rma *msg, uin
 		goto out;
 	}
 
+	/*
+	 * A handshake is required to choose the correct protocol (whether to use device read).
+	 */
+	if (!(peer->flags & EFA_RDM_PEER_HANDSHAKE_RECEIVED)) {
+		err = rxr_pkt_trigger_handshake(efa_rdm_ep, txe->addr, peer);
+		err = err ? err : -FI_EAGAIN;
+		goto out;
+	}
+
 	use_lower_ep_read = false;
 	if (efa_both_support_rdma_read(efa_rdm_ep, peer)) {
 		/* efa_both_support_rdma_read also check domain.use_device_rdma,
 		 * so we do not check it here
 		 */
 		use_lower_ep_read = true;
 	} else if (efa_mr_is_neuron(txe->desc[0])) {
-		err = efa_rdm_ep_determine_rdma_read_support(efa_rdm_ep, txe->addr, peer);
```

Comment:
Personally I think we should just remove this `else if`. No code is reaching this code path AFAICT.